pinot-tools/src/main/java/org/apache/pinot/tools/streams/PinotRealtimeSource.java

```diff
@@ -0,0 +1,191 @@
+  public static final String KEY_OF_TOPIC_NAME = "pinot.topic.name";
+  public static final long DEFAULT_MAX_MESSAGE_PER_SECOND = Long.MAX_VALUE;
+  public static final long DEFAULT_EMPTY_SOURCE_SLEEP_MS = 10;
+  final StreamDataProducer _producer;
+  final PinotSourceGenerator _generator;
+  final String _topicName;
+  final ExecutorService _executor;
+  final Properties _properties;
+  PinotStreamRateLimiter _rateLimiter;
+  protected volatile boolean _shutdown;
+
+  /**
+   * Constructs a source by passing in a Properties file, a generator, and a producer
+   * @param settings the settings for all components passed in
+   * @param generator the generator that can create data
+   * @param producer the producer to write the generator's data into
+   */
+  public PinotRealtimeSource(Properties settings, PinotSourceGenerator generator, StreamDataProducer producer) {
+    this(settings, generator, producer, null, null);
+  }
+
+  /**
+   * Constructs a source by passing in properties file, a generator, a producer and an executor service
+   * @param settings the settings for all components passed in
+   * @param generator the generator that can create data
+   * @param producer the producer to write the generator's data into
+   * @param executor the preferred executor instead of creating a thread pool. Null for default one
+   * @param rateLimiter the specialized rate limiter for customization. Null for default guava one
+   */
+  public PinotRealtimeSource(Properties settings, PinotSourceGenerator generator, StreamDataProducer producer,
+      @Nullable ExecutorService executor, @Nullable PinotStreamRateLimiter rateLimiter) {
+    _properties = settings;
+    _producer = producer;
+    Preconditions.checkNotNull(_producer, "Producer of a stream cannot be null");
+    _generator = generator;
+    Preconditions.checkNotNull(_generator, "Generator of a stream cannot be null");
+    _executor = executor == null ? Executors.newSingleThreadExecutor() : executor;
+    _topicName = settings.getProperty(KEY_OF_TOPIC_NAME);
+    Preconditions.checkNotNull(_topicName, "Topic name needs to be set via " + KEY_OF_TOPIC_NAME);
+    _rateLimiter = rateLimiter == null ? new GuavaRateLimiter(extractMaxQps(settings)) : rateLimiter;
+  }
+
+  public void run() {
+    _executor.execute(() -> {
+      while (!_shutdown) {
+        List<RowWithKey> rows = _generator.generateRows();
+        // we expect the generator implementation to return empty rows when there is no data available
+        // as a stream, we expect data to be available all the time
+        if (rows.isEmpty()) {
+          try {
+            Thread.sleep(DEFAULT_EMPTY_SOURCE_SLEEP_MS);
+          } catch (InterruptedException ex) {
+            LOGGER.warn("Interrupted from sleep, will check shutdown flag later", ex);
+          }
+        } else {
+          _rateLimiter.acquire(rows.size());
+          if (!_shutdown) {
+            _producer.produceKeyedBatch(_topicName, rows);
```

Comment:
since we already consolidated all the generator / producer usage all in this class. let's put the batch operation API here as a helper utils. 
```
static produceBatchRowWithKey(_producer, _topicName, rows) {
  // .. the default impl 
}
```

Reply:
The `default void produceKeyedBatch()` is introduced so we can have a batch interface too. I hope to introduce this method so if the stream producer does RPC calls per method, we can easily batch to improve throughput. (E.g., if the `_producer` is behind a microservice endpoint)

Reply:
+1. this actually is very helpful. https://github.com/apache/pinot/issues/8537 is likely due to a batch produce instability.